lib/src/global_packages.dart

```diff
@@ -427,9 +427,15 @@ To recompile executables, first run `$topLevelProgram pub global deactivate $nam
         dataError('${log.bold(name)} as globally activated requires '
             'unknown SDK "$name".');
       } else if (sdkName == 'dart') {
-        if (constraint.allows((sdk as DartSdk).version)) return;
-        dataError("${log.bold(name)} as globally activated doesn't "
-            'support Dart ${sdk.version}, try: $topLevelProgram pub global activate $name');
+        if (constraint.effectiveConstraint.allows((sdk as DartSdk).version)) {
+          return;
+        }
+        dataError('''
+${log.bold(name)} as globally activated doesn't support Dart ${sdk.version}.
+
+try:
+`$topLevelProgram pub global activate $name` to reactivate.
```

Comment:
`dart pub global activate foo` will resolve foo and check if it reaches the same resolution as current activation. If it does it will stop, otherwise it will re-activate with the new resolution.

Before this fix, `dart pub global activate foo` would reach the same resolution (same set of packages) and not update the sdk-constraint in the lockfile. Therefore a `deactivate` was needed.

With this fix, the sdk-constraint in the lockfile will be either:

* usable by Dart 3.0, so there is no need to activate at all, `global run` will just wortk
* not usable by Dart 3.0, so the existing resolution will not be compatible with 3.0, and either 
  * a new resolution is found and `activate` is enough, it will reactivate
  * a new resolution cannot be found and you cannot use the given package with 3.0.

Reply:
Got it, that sounds great!